File: src/main/java/net/easecation/ghosty/recording/entity/updated/EntityUpdatedScale.java
package net.easecation.ghosty.recording.entity.updated;

import cn.nukkit.utils.BinaryStream;
import net.easecation.ghosty.entity.SimulatedEntity;
import net.easecation.ghosty.recording.entity.EntityRecordNode;

/**
 * Created by Mulan Lin('Snake1999') on 2016/11/19 15:26.
 */
public class EntityUpdatedScale implements EntityUpdated {

    private float scale;

    @Override
    public void processTo(SimulatedEntity entity) {
        entity.setScale(scale);
    }

    @Override
    public EntityRecordNode applyTo(EntityRecordNode node) {
        node.setScale(scale);
        return node;
    }

    public static EntityUpdatedScale of(float scale) {
        return new EntityUpdatedScale(scale);
    }

    @Override
    public int getUpdateTypeId() {
        return TYPE_SCALE;
    }

    @Override
    public boolean hasStates() {
        return true;
    }

    public EntityUpdatedScale(BinaryStream stream) {
        this.read(stream);
    }

    private EntityUpdatedScale(float scale) {
        this.scale = scale;
    }

    @Override
    public boolean equals(Object obj) {
        if(!(obj instanceof EntityUpdatedScale o)) return false;
        return (scale == o.scale);
    }

    @Override
    public void write(BinaryStream stream) {
        stream.putFloat((float) this.scale);
    }

    @Override
    public void read(BinaryStream stream) {
        this.scale = stream.getFloat();
    }

    @Override
    public String toString() {
        return "EntityUpdatedScale{" +
            "scale=" + scale +
            '}';
    }

}

File: src/main/java/net/easecation/ghosty/recording/entity/updated/EntityUpdatedScale.kt
package net.easecation.ghosty.recording.entity.updated

import cn.nukkit.utils.BinaryStream
import kotlinx.serialization.Serializable
import net.easecation.ghosty.entity.SimulatedEntity
import net.easecation.ghosty.recording.UpdateWithState
import net.easecation.ghosty.recording.entity.EntityRecordNode

@Serializable
data class EntityUpdatedScale(
    val scale: Float
) : EntityUpdated, UpdateWithState {
    override fun getUpdateTypeId(): Int = EntityUpdated.TYPE_SCALE
    override fun processTo(entity: SimulatedEntity) {
        entity.setScale(scale)
    }

    override fun applyTo(node: EntityRecordNode): EntityRecordNode {
        node.scale = scale
        return node
    }

    companion object {
        @JvmStatic
        fun of(scale: Float) = EntityUpdatedScale(scale)

        @JvmStatic
        val ADAPTER: UpdateAdapter<EntityUpdatedScale> = Adapter
    }

    private object Adapter : UpdateAdapter<EntityUpdatedScale> {
        override fun hasStates() = true
        override fun write(updated: EntityUpdatedScale, stream: BinaryStream) {
            stream.putFloat(updated.scale)
        }
        override fun read(stream: BinaryStream) = EntityUpdatedScale(
            scale = stream.float
        )
    }
}
